desc/protoparse/parser_test.go

```diff
@@ -337,3 +340,48 @@ message Foo {
 	comment := fds[0].GetMessageTypes()[0].GetFields()[0].GetSourceInfo().GetLeadingComments()
 	testutil.Eq(t, " leading comments\n", comment)
 }
+
+func TestParseCustomOptions(t *testing.T) {
+	accessor := FileContentsFromMap(map[string]string{
+		"test.proto": `
+syntax = "proto3";
+import "google/protobuf/descriptor.proto";
+extend google.protobuf.MessageOptions {
+    string foo = 30303;
+    int64 bar = 30304;
+}
+message Foo {
+  option (.foo) = "foo";
+  option (bar) = 123;
+}
+`,
+	})
+
+	p := Parser{
+		Accessor:              accessor,
+		IncludeSourceCodeInfo: true,
+	}
+	fds, err := p.ParseFiles("test.proto")
+	testutil.Ok(t, err)
+
+	md := fds[0].GetMessageTypes()[0]
+	opts := md.GetMessageOptions()
+	data := internal.GetUnrecognized(opts)
+	buf := codec.NewBuffer(data)
+
+	tag, wt, err := buf.DecodeTagAndWireType()
```

Comment:
This repros https://github.com/bufbuild/buf/issues/52 in the `protoparse` tests. Without the fix to the `dynamic` package in this pull request, this operation fails with an EOF since there are no unrecognized bytes in the message options.